DTLS 1.3: allow fragmenting the second ClientHello message

julek-wolfssl opened the pull request:
- DTLS 1.3 pqc support
- Add `--enable-dtls-frag-ch` option to enable CH fragmenting
- Send an alert when we get an empty keyshare with a cookie present to not allow for multiple HRR in one connection
- Only update the DTLS window when we are in stateful mode
- Call `ssl->chGoodCb` as soon as we have processed a verified full or fragmented ClientHello cookie

julek-wolfssl committed:
DTLS 1.3: allow fragmenting the second ClientHello message

- DTLS 1.3 pqc support
- Add --enable-dtls-frag-ch option to enable CH fragmenting
- Send an alert when we get an empty keyshare with a cookie present to not allow for multiple HRR in one connection
- Only update the DTLS window when we have successfully processed or stored a message
- Call ssl->chGoodCb as soon as we have processed a verified full or fragmented ClientHello cookie

julek-wolfssl commented on src/tls13.c:
```diff
@@ -4410,10 +4410,48 @@ int SendTls13ClientHello(WOLFSSL* ssl)
+        int maxFrag = wolfSSL_GetMaxFragSize(ssl, MAX_RECORD_SIZE);
+        word16 lenWithoutExts = args->length;
+#endif
+
+        /* Include length of TLS extensions. */
+        ret = TLSX_GetRequestSize(ssl, client_hello, &args->length);
+        if (ret != 0)
+            return ret;
+
+#ifdef WOLFSSL_DTLS_CH_FRAG
+        if (ssl->options.dtls && args->length > maxFrag &&
+                TLSX_Find(ssl->extensions, TLSX_COOKIE) == NULL) {
+            /* Try again with an empty key share if we would be fragmenting
+             * without a cookie */
+            TLSX* ks = TLSX_Find(ssl->extensions, TLSX_KEY_SHARE);
+            if (ks == NULL) {
+                WOLFSSL_MSG("No key share and CH can't fit in one fragment.");
+                return BUFFER_ERROR;
+            }
+            args->length = lenWithoutExts;
+            if (ssl->dtls13KSE != NULL)
+                TLSX_KeyShare_FreeAll(ssl->dtls13KSE, ssl->heap);
+            ssl->dtls13KSE = (KeyShareEntry*)ks->data;
+            ks->data = NULL;
+            ret = TLSX_GetRequestSize(ssl, client_hello, &args->length);
+            if (ret != 0) {
+                /* Restore key share data */
+                ks->data = ssl->dtls13KSE;
+                ssl->dtls13KSE = NULL;
+                return ret;
+            }
+            if (args->length > maxFrag) {
+                WOLFSSL_MSG("Can't fit first CH in one fragment.");
+                return BUFFER_ERROR;
+            }
+            WOLFSSL_MSG("Sending empty key share so we don't fragment CH1");
+            ssl->options.dtlsSentEmptyKS = 1;
+        }
+#endif
+    }
```
On second thought I think it would be better to just set `ssl->options.dtlsSentEmptyKS` here and gate actually writing the key shares inside `TLSX_KeyShare_*` functions. Thoughts @rizlik ?

rizlik replied:
I would keep things way simpler by just emptying the keyshare extensions here. I would not add any specific flag.  After HRR we will re-generate only the one keyshare entry (the one selected by the server) and here we are saving all of them. We already support sending empty KeyShare in the first CH (wolfSSL_NoKeyShare) without using any extra flag. While this is a good optimization, I would leave it for a later moment.

julek-wolfssl replied:
I think I agree. Clearing it makes the change much simpler. I was hoping to not have to generate the keys twice as it can be a time consuming op. Will leave this as a future optimization.

rizlik commented on src/tls.c:
```diff
@@ -8770,24 +8769,50 @@ int TLSX_KeyShare_Parse(WOLFSSL* ssl, const byte* input, word16 length,
+                /* Clear out unusable key shares. */
+                ret = TLSX_KeyShare_Empty(ssl);
+                if (ret != 0)
+                    return ret;
```
can we generalize this improved behavior also when WOLFSSL_DTLS_CH_FRAG == 0? Also consider that RFC requires the Client to reply with just a single KeyShare after HRR.

julek-wolfssl replied:
No because when we receive a KeyShare in HRR then that means that all of the sent keys are invalid.

rizlik commented on src/tls13.c:
```diff
@@ -6644,12 +6695,21 @@ int DoTls13ClientHello(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
             goto exit_dch;
         }
+        if (ssl->chGoodCb != NULL && !IsSCR(ssl)) {
+            int cbret = ssl->chGoodCb(ssl, ssl->chGoodCtx);
```
can SCR influence v1.3 behaviour?

rizlik reviewed (changes requested):
This is very cool work!
Some comments:
1. Do you think that client good CB can be deferred until the handling of the complete verified client hello? This would reduce the changes and simplify the code.
2.  Do you think it's a good idea when sending the ch2 in response to the HRR,  to enforce that supportedVersion and cookie extensions are on the first fragment? 
3.  I think it may be helpful to add some documentation/comments explaining the requirements of point 2.

julek-wolfssl committed:
Clear the keyshare instead of storing it

julek-wolfssl commented:
> 1. Do you think that client good CB can be deferred until the handling of the complete verified client hello? This would reduce the changes and simplify the code.

I don't think so. We need to store state when we verify the first fragment. The client good CB is meant to signal that we are entering the stateful state and the `WOLFSSL` object should only be fed data from that specific peer. Otherwise, we will receive fragments and also all other connection inits being sent to the general listener.

> 2.  Do you think it's a good idea when sending the ch2 in response to the HRR,  to enforce that supportedVersion and cookie extensions are on the first fragment? 

Good idea. I'll add that.

> 3.  I think it may be helpful to add some documentation/comments explaining the requirements of point 2. 

Sounds good. Where do you think it is best to put this documentation?

julek-wolfssl committed:
Force DTLS 1.3 when accepting fragmented CH

julek-wolfssl commented on src/dtls.c:
```diff
@@ -26,6 +26,13 @@
+ * WOLFSSL_DTLS_CH_FRAG
+ *     Allow a server to process a fragmented second/verified (one containing a
+ *     valid cookie response) ClientHello message. The first/unverifies (one
+ *     without a cookie extension) ClientHello MUST be unfragmented so that the
```
Fixing.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -907,7 +938,13 @@ int DoClientHelloStateless(WOLFSSL* ssl, const byte* input,
+#ifdef WOLFSSL_DTLS_CH_FRAG
+        /* Don't send anything here when processing fragment */
+        if (isFirstCHFrag)
+            ret = BUFFER_ERROR;
```
Changing to `COOKIE_ERROR`.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -907,7 +938,13 @@ int DoClientHelloStateless(WOLFSSL* ssl, const byte* input,
+#ifdef WOLFSSL_DTLS_CH_FRAG
+        /* Don't send anything here when processing fragment */
+        if (isFirstCHFrag)
+            ret = BUFFER_ERROR;
```
If we error out on `isFirstCHFrag` without `WOLFSSL_DTLS_CH_FRAG` then we don't need this code it can be compiled out. Can you give a fuller example of what you had in mind?

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -1573,6 +1573,19 @@ static int Dtls13RtxSendBuffered(WOLFSSL* ssl)
+        return 1;
+#ifdef WOLFSSL_DTLS_CH_FRAG
+    if (ssl->options.side == WOLFSSL_SERVER_END && type == client_hello &&
+            ssl->options.dtls13ChFrag && ssl->options.dtlsStateful)
```
We only want to store further `client_hello` msgs if we are in stateful mode.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -1646,13 +1659,33 @@ static int _Dtls13HandshakeRecv(WOLFSSL* ssl, byte* input, word32 size,
+                isFirst && handshakeType == client_hello &&
+                DoClientHelloStateless(ssl, input + idx, fragLength, 1) == 0) {
+            /* We can save this message and continue as stateful. */
+            if (ssl->chGoodCb != NULL && !IsSCR(ssl)) {
```
No it shouldn't be relevant. Removing.

julek-wolfssl commented on src/internal.c:
```diff
@@ -17004,14 +17012,26 @@ static WC_INLINE int Dtls13UpdateWindow(WOLFSSL* ssl)
 
-    nextSeq = ssl->dtls13DecryptEpoch->nextPeerSeqNumber;
-    window = ssl->dtls13DecryptEpoch->window;
+    if (!w64Equal(ssl->keys.curEpoch64, ssl->dtls13DecryptEpoch->epochNumber)) {
```
Exactly. It happens because of key updates when moving to the next epoch.

julek-wolfssl commented on src/internal.c:
```diff
@@ -20782,6 +20783,18 @@ int ProcessReplyEx(WOLFSSL* ssl, int allowSocketErr)
+                                    return ret;
+                                }
+                                ret = Dtls13RecordRecvd(ssl);
+                                if (ret != 0) {
```
What if we change `Dtls13RecordRecvd` so that it doesn't add a new `Dtls13RecordNumber` if we already have the epoch/seq number already stored in `ssl->dtls13Rtx.seenRecords`?

rizlik replied:
It looks rather counter-intuitive to me as replay detection is done per record. You should also update the window when you have a record with just a fragment: in this case Dtls*DoHandshakeMsg is not invoked at all. What's the main drive to move it after the message? Is it because the window is updated in stateless code? can we reset it if we don't go stateful?

rizlik replied:
I think this is still a critical issue in the PR. Besides complicating/duplicating the code, with the actual approach the window will be updated just for the last record for a multi-records-single-message interaction, and multiple times for a single-record-multiple-messages type of packet.  Consider that RecordRecv must be per record for retransmission to work correctly. 

Can you consider changing as little as possible of the record window code path and resetting the window while we are in stateless mode?

julek-wolfssl commented on src/internal.c:
```diff
@@ -34165,8 +34247,15 @@ static int DoSessionTicket(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
+            #ifdef WOLFSSL_SEND_HRR_COOKIE
+                /* If the CH contains a cookie then we need to send an alert to
+                 * start from scratch. */
+                if (TLSX_Find(extensions, TLSX_COOKIE) != NULL)
```
I added a test case for this. `test_dtls_empty_keyshare_with_cookie` without this change fails (it sends a new HRR).

julek-wolfssl committed:
Address code review

rizlik commented:
> I don't think so. We need to store state when we verify the first fragment. The client good CB is meant to signal that we are entering the stateful state and the `WOLFSSL` object should only be fed data from that specific peer. Otherwise, we will receive fragments and also all other connection inits being sent to the general listener.

:+1:

rizlik commented:
> Sounds good. Where do you think it is best to put this documentation?

top of the file together with the define?

anhu committed:
Adding a post-quantum DTLS 1.3 test.

This exercises the fragmenting of ClientHello via large post-quantum key share.

./configure --enable-dtls-mtu --enable-dtls-frag-ch --enable-dtls \
            --enable-dtls13 --with-liboqs

julek-wolfssl committed:
fixup! Clear the keyshare instead of storing it

julek-wolfssl committed:
fixup! Clear the keyshare instead of storing it

julek-wolfssl committed:
Simplify the pqc keyshare handling

julek-wolfssl committed:
Dump manual memio stream with WOLFSSL_DUMP_MEMIO_STREAM

julek-wolfssl committed:
test_dtls13_frag_ch_pq: make sure kyber5 is used

julek-wolfssl committed:
Fix async

julek-wolfssl committed:
PQC TLS 1.3: test setting pqc with wolfSSL_CTX_set_groups

julek-wolfssl committed:
Fix unreachable code error

julek-wolfssl committed:
keyLog_callback: flush the descriptor to make sure it is written out

julek-wolfssl committed:
Fix linting issues

julek-wolfssl commented:
Retest this please

julek-wolfssl committed:
Update window in one place only when stateful

rizlik reviewed (dismissed):
Excellent work on this, it looks good to me!

JacobBarthelmeh reviewed (changes requested):
Spent a little time running a fuzzer on this, seems ok from a quick run. Only had the two minor comments to address.

julek-wolfssl commented on src/internal.c:
```diff
@@ -17086,7 +17100,7 @@ static WC_INLINE int Dtls13UpdateWindow(WOLFSSL* ssl)
 
         window[wordIndex] |= (1 << wordOffset);
-        return 1;
+        return 0;
```
Adding comment.

julek-wolfssl commented on src/tls13.c:
```diff
@@ -4410,10 +4410,37 @@ int SendTls13ClientHello(WOLFSSL* ssl)
-        return ret;
+    {
+#ifdef WOLFSSL_DTLS_CH_FRAG
+        int maxFrag = wolfSSL_GetMaxFragSize(ssl, MAX_RECORD_SIZE);
```
I'm not sure what you mean. The variables are declared at the beginning of a block so it should make c89 happy.

JacobBarthelmeh replied:
What caught my eye was the call to wolfSSL_GetMaxFragSize before declaring `word16 lenWithoutExts`. Maybe that is not an issue though with c89 builds.

julek-wolfssl committed:
Add comment

JacobBarthelmeh commented:
The curl github actions test failed twice. Could you take a look? It has succeeded on a different pull request that was recently updated.

julek-wolfssl commented:
The curl action is complaining about a missing dependency. I'll add it to the setup and see if it fixes it.

julek-wolfssl committed:
Add curl test dep

julek-wolfssl commented:
The curl test fail on master too (when I run the tests locally).

julek-wolfssl committed:
Run only stable curl tests

julek-wolfssl committed:
Increase curl timeout

julek-wolfssl committed:
Update openwrt script

julek-wolfssl committed:
Don't use /dev/null

JacobBarthelmeh reviewed (approved):
Thanks for looking at the curl CI!
